### DISCUSSION TITLE
is there any way I can encrypt data uploaded to s3?

### DISCUSSION THREAD

**Discussion by onetimecontributor** [io7m-com/exfilac#33] [Q&A] [answered]
I would like to keep only encrypted data in the cloud, so I'm wondering if there's a way to configure the tool to encrypt data before the upload? or are there plans to implement it? I've checked docs and "feature requests" so the chance is slim but still...
syncthing has a concept of [untrusted device](https://docs.syncthing.net/specs/untrusted.html#untrusted)
`Encryption is XChaCha20-Poly1305 and AES-SIV with a key derived from the password and folder ID using scrypt`
... so if I need to use some other devices as backup storage, I can reduce security risks by using that feature.

if there are no plans like that in the foreseeable future, then another question: assuming I will upload files taken from syncthing produced by that encryption, is there anything in the exfilac tool that will prevent me from doing it (like if triggers or upload will ignore/break)?
[section 2.3 of the tutorial](https://www.io7m.com/software/exfilac/documentation/d0e504.xhtml#id_2c736c6f-cce9-41df-89d8-b57f96b5b348) mentions "photos", but hopefully the tool is not limited to only image files (hasn't found anything on the matter in the manual, but maybe I've just overlooked it)

**Comment by io7m** [comment, marked as answer]
> I'm wondering if there's a way to configure the tool to encrypt data before the upload? or are there plans to implement it? I've checked docs and "feature requests" so the chance is slim but still...

There aren't any current plans to implement it, and I'm not entirely sure how it would be implemented safely.

> assuming I will upload files taken from syncthing produced by that encryption, is there anything in the exfilac tool that will prevent me from doing it (like if triggers or upload will ignore/break)?

`exfilac` treats all objects in the filesystem as opaque blobs. It'll blindly upload any files that are placed in any of the locations that you've specified.

I'm very skeptical about adding any sort of intelligence in to the app (including encryption) simply because Android is such an error-prone platform, and there are too many ways for this kind of thing to go wrong. When an app like this goes wrong, people's data is at risk... The philosophy of `exfilac` is focused on getting data off of devices as quickly as possible; my thinking is that any extra processing of data (such as encryption) is better handled by more capable systems once the data has been retrieved.

Have you considered running a [MinIO](https://min.io/) instance locally? I share your apparent distrust of third-party services. I use `exfilac` to upload files over [wireguard](https://www.wireguard.com/) to a `minio` server running on hardware that I control. I then have various scripts that I run on the same server that process the data. If I needed to replicate that data to untrusted third-party servers, like AWS, I would write something that pulled files from my `minio` server, encrypted them, and then uploaded them to AWS.

**Reply by onetimecontributor** [reply]
`Have you considered running a [MinIO](https://min.io/) instance locally`
atm I'm not at a level to get rid of cloud at all, due to kinda non-existing administration efforts and much lower costs and reliability in 10-years horizon. I just don't want my data to be stored in "plain text", though I acknowledge the risk that if my data is outside, one can eventually decrypt it, but I'm not yet trying to protect from that type of risks, more of "generic metadata farming" by big tech.
I'm on my way to pc-less setup with using Android desktop experience as a thin client where needed. 
Thus the only "offline" thing I'm considering atm is blu-ray as 2nd layer backup.

**Reply by onetimecontributor** [reply]
` philosophy of exfilac is focused`
appreciate that. assuming limited capacity it's the worst thing to scatter efforts. I'll consider to utilize syncthing's features for that.
